src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -0,0 +1,256 @@
+        }
+
+        public void SetValue(CoseHeaderLabel label, ReadOnlySpan<byte> value)
+        {
+            ValidateIsReadOnly();
+            ValidateHeaderValue(label, CborReaderState.ByteString, null);
+
+            var writer = new CborWriter();
+            writer.WriteByteString(value);
+            _headerParameters[label] = writer.Encode();
+        }
+
+        public void Remove(CoseHeaderLabel label)
+        {
+            ValidateIsReadOnly();
+            _headerParameters.Remove(label);
+        }
+
+        private void ValidateIsReadOnly()
+        {
+            if (_isReadOnly)
+            {
+                throw new InvalidOperationException(SR.CoseHeaderMapDecodedMapIsReadOnlyCannotSetValue);
+            }
+        }
+
+        private void ValidateHeaderValue(CoseHeaderLabel label, CborReaderState? state, ReadOnlyMemory<byte>? encodedValue)
+        {
+            if (state != null)
+            {
+                Debug.Assert(encodedValue == null);
+                if (label.LabelAsInt32 != null)
+                {
+                    Debug.Assert(label.LabelAsString == null);
+                    // all known headers are integers.
+                    ValidateKnownHeaderValue(label.LabelAsInt32.Value, state, null);
+                }
+            }
+            else
+            {
+                Debug.Assert(encodedValue != null);
+                var reader = new CborReader(encodedValue.Value);
+
+                if (label.LabelAsInt32 != null)
+                {
+                    Debug.Assert(label.LabelAsString == null);
+                    // all known headers are integers.
+                    ValidateKnownHeaderValue(label.LabelAsInt32.Value, reader.PeekState(), reader);
+                }
+                else
+                {
+                    Debug.Assert(label.LabelAsString != null);
+                    reader.SkipValue();
+                }
+
+                if (reader.BytesRemaining != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                }
+            }
+
+            static void ValidateKnownHeaderValue(int label, CborReaderState? initialState, CborReader? reader)
+            {
+                switch (label)
+                {
+                    case KnownHeaders.Alg:
+                        if (initialState != CborReaderState.NegativeInteger &&
+                            initialState != CborReaderState.UnsignedInteger &&
+                            initialState != CborReaderState.TextString)
+                        {
+                            throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                        }
+                        reader?.SkipValue();
+                        break;
+                    case KnownHeaders.Crit:
+                        throw new NotSupportedException(); // TODO
```

Comment:
There's an interesting question of whether or not this routine should run during Decode.

If it does, then any malformed/future thing can't be read at all, vs Verify fails.  If it doesn't, then Verify has more error states to deal with.

I don't know I have a strong feeling one way or the other.  My main concern is that by throwing for Crit or CounterSignature that we may make the package unusable for our early partners.  If we believe they don't need it yet, then it's reasonable.

Reply:
> My main concern is that by throwing for Crit or CounterSignature that we may make the package unusable for our early partners.

This is a very good point, we shouldn't make it unusable for those cases. I think that you should be able to decode messages carrying Critical or Counter Signature headers and the unsupported error should be moved to Verify. 
On a similar way, you should be able to Encode (via `Sign(...)`) these kind of messages.